extra/emacs/.emacs.d/config.org
#+title: config.org
#+subtitle: david's emacs config

* About
This is my fully literate emacs config.
inspiration: https://github.com/angrybacon/dotemacs
* Bootstrap
** Better defaults
When we're running in GUI mode, we want to disable cursor blinking, the scroll bar, and the toolbar.

#+begin_src emacs-lisp
(when window-system
  (blink-cursor-mode 0) ; Disable the cursor blinking
  (scroll-bar-mode 0)   ; Disable the scroll bar
  (tool-bar-mode 0)     ; Disable the tool bar
  (tooltip-mode 1))     ; Enable the tooltips
#+end_src

#+begin_src emacs-lisp
(xterm-mouse-mode 1)
#+end_src

#+begin_src emacs-lisp
(setq evil-want-keybinding nil)
(setq-default
 mac-function-modifier 'hyper

 ad-redefinition-action 'accept                   ; Silence warnings for redefinition
 auto-window-vscroll nil                          ; Lighten vertical scroll
 confirm-kill-emacs nil                           ; Confirm before exiting Emacs
 cursor-in-non-selected-windows t                 ; Hide the cursor in inactive windows
 delete-by-moving-to-trash t                      ; Delete files to trash
 display-time-default-load-average nil            ; Don't display load average
 display-time-format "%Y-%m-%d %H:%M"             ; Format the time string
 echo-keystrokes 0.001
 fill-column 80                                   ; Set width for automatic line breaks
 help-window-select t                             ; Focus new help windows when opened
 indent-tabs-mode nil                             ; Stop using tabs to indent
 inhibit-startup-screen t                         ; Disable start-up screen
 initial-major-mode 'literate-calc-mode
 initial-scratch-message ""                       ; Empty the initial *scratch* buffer
 scroll-conservatively most-positive-fixnum       ; Always scroll by one line
 scroll-margin 10                                 ; Add a margin when scrolling vertically
 select-enable-clipboard t                        ; Merge system's and Emacs' clipboard
 sentence-end-double-space nil                    ; End a sentence after a dot and a space
 show-trailing-whitespace nil                     ; Display trailing whitespaces
 split-height-threshold nil                       ; Disable vertical window splitting
 split-width-threshold nil                        ; Disable horizontal window splitting
 tab-always-indent 'complete
 tab-width 4                                      ; Set width for tabs
 uniquify-buffer-name-style 'forward              ; Uniquify buffer names
 use-dialog-box nil
 help-at-pt-display-when-idle t
 help-at-pt-timer-delay 0
 window-combination-resize t                      ; Resize windows proportionally
 x-stretch-cursor t                               ; Stretch cursor to the glyph width
 frame-title-format '((:eval (if (buffer-file-name)
                                 (abbreviate-file-name (buffer-file-name))
                               "%b"))))

(help-at-pt-set-timer)

;; Put autosaves in emacs-saves
(setq auto-save-file-name-transforms
  `((".*" "~/.emacs-saves/" t)))
(setq backup-directory-alist '(("." . "~/.emacs-saves")))

(setq max-lisp-eval-depth 5000)
(setq max-specpdl-size 5000)

(setq org-startup-indented t)

(setq mouse-autoselect-window t
      focus-follows-mouse t)

(cd "~/")                                         ; Move to the user directory
(delete-selection-mode 1)                         ; Replace region when inserting text
; (display-time-mode 1)                             ; Enable time in the mode-line
(fset 'yes-or-no-p 'y-or-n-p)                     ; Replace yes/no prompts with y/n
(global-subword-mode 1)                           ; Iterate through CamelCase words
; https://github.com/railwaycat/homebrew-emacsmacport/issues/139
(menu-bar-mode 1)
(put 'downcase-region 'disabled nil)              ; Enable downcase-region
(put 'upcase-region 'disabled nil)                ; Enable upcase-region
(set-default-coding-systems 'utf-8)               ; Default to utf-8 encoding
(column-number-mode)                              ; Show col nums in modeline
#+end_src
*** Backups
#+begin_src emacs-lisp
(setq backup-directory-alist `(("." . "~/.saves")))
(setq backup-by-copying t)
#+end_src
** Custom file
#+begin_src emacs-lisp
(setq-default custom-file (expand-file-name ".custom.el" user-emacs-directory))
(when (file-exists-p custom-file)
  (load custom-file))
#+end_src
** Miscellaneous delights
#+begin_src emacs-lisp
(delight '((subword-mode nil t)
           (eldoc-mode nil t)))
#+end_src
** Evil
#+begin_src emacs-lisp
(use-package undo-fu)

(use-package evil
  :custom
  (evil-undo-system 'undo-fu)
  (evil-want-keybindings nil)
  :init
  (setq
   evil-want-C-u-scroll t
   evil-undo-system 'undo-fu
   evil-want-C-i-jump nil)
  (evil-mode)
  (evil-normalize-keymaps)
  (setq-default evil-undo-system 'undo-fu)
  :config
  (define-key evil-normal-state-map "u" 'undo-fu-only-undo)
  (define-key evil-normal-state-map "\C-r" 'undo-fu-only-redo))

(add-hook 'window-configuration-change-hook #'evil-normalize-keymaps)
(add-hook 'org-src-mode-hook #'evil-normalize-keymaps)

(use-package evil-collection
  :after evil
  :init
  (setq evil-want-keybindings nil)
  :config
  (evil-collection-init '(magit notmuch calendar dired pdf)))
#+end_src
*** Evil Fringe Mark
o show marks in the left fringe of the editor.
#+begin_src emacs-lisp
(use-package evil-fringe-mark
 :delight global-evil-fringe-mark-mode
 :config
 (global-evil-fringe-mark-mode))
#+end_src
*** evil-surround
#+begin_src emacs-lisp
(use-package evil-surround
  :config
  (global-evil-surround-mode 1))
#+end_src
** general.el
*** Leader definitions
#+begin_src emacs-lisp
(defconst dcao/leader-key "SPC")
(defconst dcao/leader-key-alt "C-SPC")
(defconst dcao/local-leader-key "SPC m")
(defconst dcao/local-leader-key-alt "C-SPC m")
#+end_src
*** Bind configuration
#+begin_src emacs-lisp
(use-package general
  :after evil
  :init
  (setq-default evil-undo-system 'undo-tree)
  :config
  (general-evil-setup t)
  (general-unbind 'normal dcao/leader-key)
  (general-unbind 'emacs dcao/leader-key)
  (general-unbind dcao/leader-key-alt)
  
  (general-create-definer dcao-leader-def
    :prefix dcao/leader-key-alt
    :keymaps 'override)
  (general-create-definer dcao-local-leader-def
    :prefix dcao/local-leader-key-alt
    :keymaps 'override)
  
  (general-def '(normal visual) "SPC" (general-simulate-key "C-SPC"))

  (general-def "<escape>" 'keyboard-escape-quit)
  (general-def "C-w" 'backward-kill-word)

  (general-def "M-w" 'kill-this-buffer)
  (general-def "M-q" 'delete-window)

  (general-def "M-P" 'execute-extended-command)

  (general-def "M-[" 'evil-jump-backward)
  (general-def "M-]" 'evil-jump-forward)

  (general-def
    :keymaps 'helm-map
    "C-w" 'backward-kill-word)

  (general-def "C-s" 'save-buffer)

  (general-def dcao/leader-key-alt (general-simulate-key dcao/leader-key))

  (dcao-leader-def
    ;; unbind SPC and give it a title for which-key (see echo area)
    "" '(nil :which-key "universal leader")
    "u" 'universal-argument
    ;; bind nothing but give SPC f a description for which-key
    "f" '(:ignore t :which-key "file")
    ;; use a cons as a replacement
    "g" '(:ignore t :which-key ("g-key" . "git"))
    "x" '(:ignore t :which-key "meta")
    "b k" 'kill-buffer
    "b q" 'kill-this-buffer
    "w q" 'delete-window
    "w /" 'evil-window-vsplit
    "w -" 'evil-window-split
    "w h" 'evil-window-left
    "w j" 'evil-window-down
    "w k" 'evil-window-up
    "w l" 'evil-window-right
    "w n" 'make-frame-command
    "x r" 'dcao/reload-config
    "x q" 'save-buffers-kill-terminal
    "c w" 'count-words
    "c r" 'count-words-region)

  (general-define-key
   :keymaps 'override
   dcao/local-leader-key-alt '(nil :which-key "local leader")))
#+end_src
*** evil-numbers
#+begin_src emacs-lisp
(use-package evil-numbers
  :general
  ("C-a" 'evil-numbers/inc-at-pt))
#+end_src
** buffer-move
#+begin_src emacs-lisp
;;; buffer-move.el --- 

;; Copyright (C) 2004-2014  Lucas Bonnet <lucas@rincevent.net.fr>

;; Author: Lucas Bonnet <lucas@rincevent.net>
;; Keywords: lisp,convenience
;; Version: 0.5
;; URL : https://github.com/lukhas/buffer-move

;; This program is free software; you can redistribute it and/or
;; modify it under the terms of the GNU General Public License
;; as published by the Free Software Foundation; either version 2
;; of the License, or (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program; if not, write to the Free Software
;; Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
;; 02111-1307, USA.

;;; Commentary:

;; This file is for lazy people wanting to swap buffers without
;; typing C-x b on each window. This is useful when you have :

;; +--------------+-------------+
;; |              |             |
;; |    #emacs    |    #gnus    |
;; |              |             |
;; +--------------+-------------+
;; |                            |
;; |           .emacs           |
;; |                            |
;; +----------------------------+

;; and you want to have :

;; +--------------+-------------+
;; |              |             |
;; |    #gnus     |   .emacs    |
;; |              |             |
;; +--------------+-------------+
;; |                            |
;; |           #emacs           |
;; |                            |
;; +----------------------------+

;; With buffer-move, just go in #gnus, do buf-move-left, go to #emacs
;; (which now should be on top right) and do buf-move-down.

;; To use it, simply put a (require 'buffer-move) in your ~/.emacs and
;; define some keybindings. For example, i use :

;; (global-set-key (kbd "<C-S-up>")     'buf-move-up)
;; (global-set-key (kbd "<C-S-down>")   'buf-move-down)
;; (global-set-key (kbd "<C-S-left>")   'buf-move-left)
;; (global-set-key (kbd "<C-S-right>")  'buf-move-right)


;;; Code:


(require 'windmove)

;;;###autoload
(defun buf-move-up ()
  "Swap the current buffer and the buffer above the split.
If there is no split, ie now window above the current one, an
error is signaled."
;;  "Switches between the current buffer, and the buffer above the
;;  split, if possible."
  (interactive)
  (let* ((other-win (windmove-find-other-window 'up))
	 (buf-this-buf (window-buffer (selected-window))))
    (if (null other-win)
        (error "No window above this one")
      ;; swap top with this one
      (set-window-buffer (selected-window) (window-buffer other-win))
      ;; move this one to top
      (set-window-buffer other-win buf-this-buf)
      (select-window other-win))))

;;;###autoload
(defun buf-move-down ()
"Swap the current buffer and the buffer under the split.
If there is no split, ie now window under the current one, an
error is signaled."
  (interactive)
  (let* ((other-win (windmove-find-other-window 'down))
	 (buf-this-buf (window-buffer (selected-window))))
    (if (or (null other-win) 
            (string-match "^ \\*Minibuf" (buffer-name (window-buffer other-win))))
        (error "No window under this one")
      ;; swap top with this one
      (set-window-buffer (selected-window) (window-buffer other-win))
      ;; move this one to top
      (set-window-buffer other-win buf-this-buf)
      (select-window other-win))))

;;;###autoload
(defun buf-move-left ()
"Swap the current buffer and the buffer on the left of the split.
If there is no split, ie now window on the left of the current
one, an error is signaled."
  (interactive)
  (let* ((other-win (windmove-find-other-window 'left))
	 (buf-this-buf (window-buffer (selected-window))))
    (if (null other-win)
        (error "No left split")
      ;; swap top with this one
      (set-window-buffer (selected-window) (window-buffer other-win))
      ;; move this one to top
      (set-window-buffer other-win buf-this-buf)
      (select-window other-win))))

;;;###autoload
(defun buf-move-right ()
"Swap the current buffer and the buffer on the right of the split.
If there is no split, ie now window on the right of the current
one, an error is signaled."
  (interactive)
  (let* ((other-win (windmove-find-other-window 'right))
	 (buf-this-buf (window-buffer (selected-window))))
    (if (null other-win)
        (error "No right split")
      ;; swap top with this one
      (set-window-buffer (selected-window) (window-buffer other-win))
      ;; move this one to top
      (set-window-buffer other-win buf-this-buf)
      (select-window other-win))))


(provide 'buffer-move)
;;; buffer-move.el ends here
#+end_src
** gcmh
The Garbage Collector Magic Hack delays emacs' GC from collecting until idle.

#+begin_src emacs-lisp
(use-package gcmh
  :delight
  :config
  (gcmh-mode 1))
#+end_src
** Meta
*** Config reload
#+begin_src emacs-lisp
(defun dcao/reload-config ()
  (interactive)
  (load-file (expand-file-name "init.el" user-emacs-directory)))
#+end_src
*** Config compilation
#+begin_src emacs-lisp
(defun dcao/compile-config ()
  (interactive)
  (byte-compile-file (expand-file-name "config.el" user-emacs-directory))
  (byte-compile-file (expand-file-name "init.el" user-emacs-directory)))
#+end_src
** init shell
#+begin_src emacs-lisp
(use-package exec-path-from-shell
  :init
  (setq exec-path-from-shell-arguments nil)
  (when (memq window-system '(mac ns x))
    (exec-path-from-shell-initialize)))
#+end_src
** recentf
#+begin_src emacs-lisp
(use-package recentf
  :init
  (setq recentf-max-saved-items 1000
        recentf-auto-cleanup 'never
        recentf-auto-save-timer (run-with-idle-timer 600 t
                                                     'recentf-save-list))

  :config
  (recentf-mode 1))
#+end_src
** delete this file
#+begin_src emacs-lisp
(defun delete-file-and-buffer ()
  "Kill the current buffer and deletes the file it is visiting."
  (interactive)
  (let ((filename (buffer-file-name)))
    (if filename
        (if (y-or-n-p (concat "Do you really want to delete file " filename " ?"))
            (progn
              (delete-file filename)
              (message "Deleted file %s." filename)
              (kill-buffer)))
      (message "Not a file visiting buffer!"))))

(dcao-leader-def
  "f d" 'delete-file-and-buffer)
#+end_src
** moving between files
#+begin_src emacs-lisp
(general-define-key "<A-M-right>" 'next-buffer
                    "<A-M-left>"  'previous-buffer)
#+end_src
* UI
** exwm
#+begin_src emacs-lisp
(defun dcao/exwm-launch (command)
  (interactive)
  (let ((default-directory (projectile-project-root)))
    (start-process-shell-command command nil command)))
    
(defun dcao/exwm-bind-launch (command)
  (interactive)
  `(lambda () (interactive) (dcao/exwm-launch ,command)))

(defun dcao/exwm-ws-swap-mon (ws &optional switch force)
  (let ((mon (plist-get exwm-randr-workspace-output-plist ws)))
    (when (or force (not (equal exwm-workspace-current-index ws)))
      (if (string= mon "eDP1")
          (progn
            (plist-put exwm-randr-workspace-output-plist ws "HDMI2")
            (exwm-randr-refresh))
      (plist-put exwm-randr-workspace-output-plist ws "eDP1")
      (exwm-randr-refresh))))
  (when (switch)
    (exwm-workspace-switch ws)))
#+end_src

#+begin_src emacs-lisp
(use-package exwm
  :general
  (:keymaps 'exwm-mode-map
    "M-<tab>" 'consult-buffer
    "M-l" (lambda () (interactive) (split-window nil nil 'right))
    "M-h" (lambda () (interactive) (split-window nil nil 'left))
    "M-j" 'split-window-below)
  :custom
  (exwm-input-prefix-keys '(?\C-\s ?\M-x ?\C-\M-p ?\C-\M-f ?\M-\t))
  :config
  (use-package exwm-config :straight nil)
  ; (use-package exwm-randr :straight nil)

  ; (setq exwm-randr-workspace-output-plist
  ;       '(0 "eDP1"
  ;         1 "HDMI2"
  ;         2 "HDMI2"
  ;         3 "HDMI2"
  ;         4 "eDP1"
  ;         5 "eDP1"
  ;         6 "eDP1"
  ;         7 "eDP1"
  ;         8 "eDP1"
  ;         9 "eDP1"))
  ; (add-hook 'exwm-randr-screen-change-hook
  ;           (lambda ()
  ;             (start-process-shell-command
  ;              "xrandr" nil "xrandr --output HDMI2 --right-of eDP1 --auto")))
  ; (exwm-randr-enable)

  (setq
   exwm-workspace-number 9
   exwm-workspace-show-all-buffers t
   exwm-layout-show-all-buffers t)

  (display-time-mode t)
  (display-battery-mode t)
  
  (use-package desktop-environment
    
    :config
    (setq desktop-environment-brightness-get-command "light")
    (setq desktop-environment-brightness-set-command "light %s")
    (setq desktop-environment-brightness-get-regexp "^\\([0-9]+\\)")
    (setq desktop-environment-brightness-normal-increment "-A 5")
    (setq desktop-environment-brightness-normal-decrement "-U 5")
    (setq desktop-environment-brightness-small-increment "-A 5")
    (setq desktop-environment-brightness-small-decrement "-U 5")
    
    (exwm-input-set-key (kbd "<XF86AudioMute>") #'desktop-environment-toggle-mute)
    (exwm-input-set-key (kbd "<XF86AudioMicMute>") #'desktop-environment-toggle-microphone-mute)
    (exwm-input-set-key (kbd "<XF86AudioRaiseVolume>") #'desktop-environment-volume-increment)
    (exwm-input-set-key (kbd "<XF86AudioLowerVolume>") #'desktop-environment-volume-decrement)
    (exwm-input-set-key (kbd "<XF86MonBrightnessUp>") #'desktop-environment-brightness-increment)
    (exwm-input-set-key (kbd "<XF86MonBrightnessDown>") #'desktop-environment-brightness-decrement)
    (exwm-input-set-key (kbd "<XF86Bluetooth>") #'desktop-environment-toggle-bluetooth)

    (require 'cl-lib)
    (dolist (k '(XF86AudioMute
                 XF86AudioMicMute
                 XF86AudioRaiseVolume
                 XF86AudioLowerVolume
                 XF86MonBrightnessUp
                 XF86MonBrightnessDown
                 XF86Bluetooth))
      (pushnew k exwm-input-prefix-keys)))

  (add-hook 'exwm-update-class-hook
            (lambda ()
              (unless (or (string-prefix-p "sun-awt-X11-" exwm-instance-name)
                          (string= "gimp" exwm-instance-name))
                (exwm-workspace-rename-buffer exwm-class-name))))
              
  (add-hook 'exwm-update-title-hook
            (lambda ()
              (when (or (not exwm-instance-name)
                        (string-prefix-p "sun-awt-X11-" exwm-instance-name)
                        (string= "gimp" exwm-instance-name))
                (exwm-workspace-rename-buffer exwm-title))))

  (exwm-input-set-key (kbd "<s-tab>") #'helm-run-external-command)
  (exwm-input-set-key (kbd "<M-tab>") #'consult-buffer)
  (exwm-input-set-key (kbd "s-h") #'evil-window-left)
  (exwm-input-set-key (kbd "s-j") #'evil-window-down)
  (exwm-input-set-key (kbd "s-k") #'evil-window-up)
  (exwm-input-set-key (kbd "s-l") #'evil-window-right)
  (exwm-input-set-key (kbd "s-w") #'kill-this-buffer)
  (exwm-input-set-key (kbd "s-q") #'delete-window)

  (exwm-input-set-key (kbd "s-C-h") #'shrink-window-horizontally)
  (exwm-input-set-key (kbd "s-C-l") #'enlarge-window-horizontally)
  (exwm-input-set-key (kbd "s-C-j") #'shrink-window)
  (exwm-input-set-key (kbd "s-C-k") #'enlarge-window)
  
  (exwm-input-set-key (kbd "s-K") #'buf-move-up)
  (exwm-input-set-key (kbd "s-J") #'buf-move-down)
  (exwm-input-set-key (kbd "s-H") #'buf-move-left)
  (exwm-input-set-key (kbd "s-L") #'buf-move-right)
  
  (exwm-input-set-key (kbd "M-l") (lambda () (interactive) (split-window nil nil 'right)))
  (exwm-input-set-key (kbd "M-h") (lambda () (interactive) (split-window nil nil 'left)))
  (exwm-input-set-key (kbd "M-j") #'split-window-below)

  (dolist (k '(?\s-H
               ?\s-J
               ?\s-K
               ?\C-\s
               ?\s-L))
    (pushnew k exwm-input-prefix-keys))
  
  (exwm-input-set-key (kbd "<s-return>")
                      (lambda ()
                        (interactive)
                        (start-process-shell-command "st" nil
                                                     (if (projectile-project-root)
                                                         (concat "sh -c 'cd " (projectile-project-root) "; st'")
                                                       "st"))))

  (exwm-input-set-key (kbd "C-M-p")
                      (lambda () (interactive) (start-process-shell-command "rofi-pass" nil "rofi-pass")))
  (exwm-input-set-key (kbd "C-M-f")
                      (lambda () (interactive) (start-process-shell-command "flameshot" nil "flameshot gui")))

  (exwm-input-set-key (kbd "s-1")
                      (lambda () (interactive) (exwm-workspace-switch 0)))
  (exwm-input-set-key (kbd "s-2")
                      (lambda () (interactive) (exwm-workspace-switch 1)))
  (exwm-input-set-key (kbd "s-3")
                      (lambda () (interactive) (exwm-workspace-switch 2)))
  (exwm-input-set-key (kbd "s-4")
                      (lambda () (interactive) (exwm-workspace-switch 3)))
  (exwm-input-set-key (kbd "s-5")
                      (lambda () (interactive) (exwm-workspace-switch 4)))
  (exwm-input-set-key (kbd "s-6")
                      (lambda () (interactive) (exwm-workspace-switch 5)))
  (exwm-input-set-key (kbd "s-7")
                      (lambda () (interactive) (exwm-workspace-switch 6)))
  (exwm-input-set-key (kbd "s-8")
                      (lambda () (interactive) (exwm-workspace-switch 7)))
  (exwm-input-set-key (kbd "s-9")
                      (lambda () (interactive) (exwm-workspace-switch 8)))
  (exwm-input-set-key (kbd "s-0")
                      (lambda () (interactive) (exwm-workspace-switch 9)))
                      
  (exwm-input-set-key (kbd "s-!")
                      (lambda () (interactive) (exwm-workspace-move-window 0)))
  (exwm-input-set-key (kbd "s-@")
                      (lambda () (interactive) (exwm-workspace-move-window 1)))
  (exwm-input-set-key (kbd "s-#")
                      (lambda () (interactive) (exwm-workspace-move-window 2)))
  (exwm-input-set-key (kbd "s-$")
                      (lambda () (interactive) (exwm-workspace-move-window 3)))
  (exwm-input-set-key (kbd "s-%")
                      (lambda () (interactive) (exwm-workspace-move-window 4)))
  (exwm-input-set-key (kbd "s-^")
                      (lambda () (interactive) (exwm-workspace-move-window 5)))
  (exwm-input-set-key (kbd "s-&")
                      (lambda () (interactive) (exwm-workspace-move-window 6)))
  (exwm-input-set-key (kbd "s-*")
                      (lambda () (interactive) (exwm-workspace-move-window 7)))
  (exwm-input-set-key (kbd "s-(")
                      (lambda () (interactive) (exwm-workspace-move-window 8)))
  (exwm-input-set-key (kbd "s-)")
                      (lambda () (interactive) (exwm-workspace-move-window 9)))
  
  
  
  (dolist (k '(?\C-\s-!
               ?\C-\s-@
               ?\C-\s-#
               ?\C-\s-$
               ?\C-\s-%
               ?\C-\s-^
               ?\C-\s-&
               ?\C-\s-*
               ?\C-\s-\(
               ?\C-\s-\)))
    (pushnew k exwm-input-prefix-keys))
  
  (exwm-input-set-key (kbd "s-C-!")
                      (lambda () (interactive) (dcao/exwm-ws-swap-mon 0)))
  (exwm-input-set-key (kbd "s-C-@")
                      (lambda () (interactive) (dcao/exwm-ws-swap-mon 1)))
  (exwm-input-set-key (kbd "s-C-#")
                      (lambda () (interactive) (dcao/exwm-ws-swap-mon 2)))
  (exwm-input-set-key (kbd "s-C-$")
                      (lambda () (interactive) (dcao/exwm-ws-swap-mon 3)))
  (exwm-input-set-key (kbd "s-C-%")
                      (lambda () (interactive) (dcao/exwm-ws-swap-mon 4)))
  (exwm-input-set-key (kbd "s-C-^")
                      (lambda () (interactive) (dcao/exwm-ws-swap-mon 5)))
  (exwm-input-set-key (kbd "s-C-&")
                      (lambda () (interactive) (dcao/exwm-ws-swap-mon 6)))
  (exwm-input-set-key (kbd "s-C-*")
                      (lambda () (interactive) (dcao/exwm-ws-swap-mon 7)))
  (exwm-input-set-key (kbd "s-C-(")
                      (lambda () (interactive) (dcao/exwm-ws-swap-mon 8)))
  (exwm-input-set-key (kbd "s-C-)")
                      (lambda () (interactive) (dcao/exwm-ws-swap-mon 9)))

  (require 'exwm-systemtray)
  (exwm-systemtray-enable)

  (exwm-init))
#+end_src
** Theme
#+begin_src emacs-lisp
(eval-and-compile
  (defun dcao/lisp-dir ()
    (expand-file-name "lisp" user-emacs-directory)))

(setq
 dcao/default-font (font-spec :family "Iosevka Custom" :size 16)
 dcao/mono-font (font-spec :family "Iosevka Custom" :size 16)
 dcao/variable-font (font-spec :family "iA Writer Quattro S" :size 14 :weight 'ultra-light))

(set-face-attribute 'default nil :font dcao/default-font)
(set-face-attribute 'fixed-pitch nil :font dcao/mono-font)
(set-face-attribute 'variable-pitch nil :font dcao/variable-font)

; (use-package nano-theme
;   :straight (:type git :host github
;              :repo "rougier/nano-theme")
;   :config
;   (load-theme 'nano-dark t))

(use-package doom-themes
  :config
  (load-theme 'doom-gruvbox t)
  (doom-themes-org-config))
#+end_src
** Modeline
#+begin_src emacs-lisp
(use-package hide-mode-line)
#+end_src

#+begin_src emacs-lisp

(use-package simple-modeline
  :hook (after-init . simple-modeline-mode)
  :config
  (setq simple-modeline-word-count-modes '()
        simple-modeline-segments '((simple-modeline-segment-modified
                                    simple-modeline-segment-buffer-name
                                    simple-modeline-segment-position)
                                   (simple-modeline-segment-input-method

extra/emacs/.doom.d/config.el
;;; config/default/config.el -*- lexical-binding: t; -*-

(setq doom-font (font-spec :family "Iosevka" :size 16))
; (setq doom-font (font-spec :family "InputMonoCompressed Medium" :size 19))
(setq doom-variable-pitch-font (font-spec :family "Libre Baskerville" :size 14))
; (setq doom-variable-pitch-font (font-spec :family "Input Serif Compressed Medium" :size 18))
(setq doom-theme 'doom-gruvbox)
;; doom-unicode-font
(setq display-line-numbers-width 4)
(setq-default line-spacing 4)

(set-frame-parameter nil 'internal-border-width 18)

(setq initial-major-mode 'org-mode)
(setq confirm-kill-emacs nil)
(xterm-mouse-mode 1)

;;
;; Config
;;

(after! cc-mode
  (c-add-style
   "dcao"
   '((c-basic-offset . 4)))
  (setq c-default-style "dcao"))

;; Rust
(after! lsp-mode
  (add-hook! 'rust-mode-hook 'lsp-mode))

;; Magit
(def-package! forge
  :after magit
  :bind
  (:map magit-mode-map
    ("`" . forge-dispatch)))

;; Org wrangling
(defvar cao/org-root (concat (getenv "HOME") "/default/org/"))

(defvar cao/org-inbox-template "* TODO %^{Task}
:PROPERTIES:
:CREATED: %U
:END:
%i")

(defvar cao/org-contact-template "* %^{Name}
:PROPERTIES:
:BIRTHDAY: %^{DOB (yyyy-mm-dd)}
:END:
%i")

(defvar cao/org-song-rec-template "** %^{Name}
:PROPERTIES:
:CREATED: %U
:END:
%i")

(defvar cao/org-weekly-review-template "** %(format-time-string \"%Y-%V\")
:PROPERTIES:
:CREATED: %U
:END:
- [ ] Sift inbox
- [ ] Task checkup
  - [ ] Emails?
- [ ] =lt= checkup
- [ ] Self-eval
%?")

(if (featurep! :completion ivy)
  (after! ivy
    ; (setq ivy-re-builders-alist ((t . ivy--regex-plus)))
    (ivy-rich-mode +1)))

(def-package! ox-tufte
  :after org)

(after! elfeed
  (def-package! elfeed-web
    :commands elfeed-web-start
    :config
    (setq httpd-port 3749))

  (setq shr-inhibit-images t))

(after! org
  (def-package! org-contacts)
  (def-package! org-expiry)

  (setq org-agenda-files `(,cao/org-root)
        org-directory cao/org-root
        org-archive-location (concat cao/org-root "archive/%s::")
        org-log-done 'time
        org-log-into-drawer t
        org-expiry-inactive-timestamps t
        org-default-priority ?C
        org-lowest-priority ?D
        org-catch-invisible-edits 'show-and-error
        ;; refile
        org-refile-targets '((org-agenda-files :maxlevel . 3))
        org-refile-use-outline-path 'file
        org-outline-path-complete-in-steps nil
        org-refile-allow-creating-parent-nodes 'confirm
        ;; contacts
        org-contacts-files `(,(concat cao/org-root "ppl.org"))
        ;; capture
        org-capture-templates
        `(("t" "inbox todo" entry (file ,(concat cao/org-root "inbox.org"))
           ,cao/org-inbox-template)
          ("c" "contact" entry (file ,(concat cao/org-root "inbox.org"))
           ,cao/org-contact-template)
          ("s" "song rec" entry (file+headline ,(concat cao/org-root "lt.org") "Song rec")
           ,cao/org-song-rec-template)
          ("r" "weekly review" entry (file+headline ,(concat cao/org-root "review.org") ,(format-time-string "%Y"))
           ,cao/org-weekly-review-template)))

  (def-package! org-journal
    :defer t
    :custom
    (org-journal-dir (concat cao/org-root "journal/"))
    (org-journal-file-type 'yearly)
    (org-journal-date-format "%a, %B %d, %Y"))

  (setq org-agenda-custom-commands
        '(("t" "test" todo "DONE"
           ((org-agenda-prefix-format "%t k")
            (org-agenda-overriding-header "\ntest\n")))))

  (require 'appt)
  (require 'notifications)
  (appt-activate t)

  (setq appt-message-warning-time 5) ; Show notification 5 minutes before event
  (setq appt-display-interval appt-message-warning-time) ; Disable multiple reminders
  (setq appt-display-mode-line nil)

  ; Use appointment data from org-mode
  (defun cao/build-notifs ()
    (interactive)
    (setq appt-time-msg-list nil)
    (org-agenda-to-appt t nil :deadline :scheduled :timestamp :sexp))

  ; Alarm updating
  ; If Emacs is open at 12am, this will update the notifications then. Otherwise,
  ; it'll update them now.
  ; (run-at-time "12:00am" (* 24 3600) 'cao/build-notifs)
  (add-hook 'after-save-hook
            '(lambda ()
               (if (string-prefix-p cao/org-root (buffer-file-name))
                   (cao/build-notifs))))

  (add-hook 'org-agenda-finalize-hook
            'cao/build-notifs)

  (defun cao/appt-display (min-to-app new-time msg)
    (if (atom min-to-app)
      (notifications-notify :title (concat "org +" min-to-app " min") :body msg)
    (dolist (i (number-sequence 0 (1- (length min-to-app))))
      (notifications-notify :title (concat "org +" (nth i min-to-app) " min") :body (nth i msg)))))

  ; Display appointments as a window manager notification
  (setq appt-disp-window-function 'cao/appt-display)
  (setq appt-delete-window-function (lambda () t))

  (def-package! helm
    ;; If the helm feature is enabled, we don't need to redeclare it
    :unless (featurep! :completion helm)
    :defer t)

  (def-package! helm-org-rifle
    :after (org helm)
    :commands helm-org-rifle-agenda-files
    :config
    (helm-autoresize-mode +1))

  (map! :map org-mode-map
        :ni "<M-right>" #'org-narrow-to-subtree
        :ni "<M-left>" #'widen
        :localleader
        :desc "archive"     :n "a" #'org-archive-subtree
        :desc "add created" :n "e" #'org-expiry-insert-created
        :desc "refile"      :n "r" #'org-refile
        :desc "narrow"      :n "n" #'org-narrow-to-subtree
        :desc "schedule"    :n "s" #'org-schedule
        :desc "widen"       :n "w" #'widen
        :desc "export"      :n "x" #'org-export-dispatch
        :desc "timestamp"   :n "m" #'org-time-stamp
        :desc "set tags"    :n "g" #'org-set-tags-command
        :desc "set prio"    :n "p" #'org-priority))

(def-package! ripgrep)
(def-package! helm-rg)

;; Calc
(after! calc
  (setq calc-algebraic-mode t))

(defun cao/calendar ()
  "Activate (or switch to) `calendar' in its workspace."
  (interactive)
  (if (featurep! :ui workspaces)
      (progn
        (+workspace-switch "Calendar" t)
        (doom/open-scratch-buffer t)
        (+calendar/open-calendar)
        (+workspace/display))
    (setq +calendar--wconf (current-window-configuration))
    (delete-other-windows)
    (doom/open-scratch-buffer t)
    (+calendar/open-calendar)))

;;
;; Bindings
;;

(map! :leader
      (:desc "buffer" :prefix "b"
        :desc "Switch workspace buffer" :n "B" #'persp-switch-to-buffer
        :desc "Switch buffer"           :n "b" #'switch-to-buffer)
      (:desc "file" :prefix "f"
        :desc "Find file"               :n "f" #'find-file
        :desc "rg project"              :n "z" #'helm-projectile-rg
        :desc "Org dir"                 :n "o" (lambda () (interactive) (helm-find-files-1 cao/org-root)))
      (:desc "org" :prefix "o"
        :desc "Capture"    :n "c" #'org-capture
        :desc "Calendar"   :n "C" #'cao/calendar
        :desc "Mail"       :n "m" #'notmuch
        :desc "journal"    :n "j" #'org-journal-new-entry
        :desc "Org dir"    :n "f" (lambda () (interactive) (helm-find-files-1 cao/org-root))
        :desc "Org rifle"  :n "r" #'helm-org-rifle-agenda-files)
      (:desc "window" :prefix "w"
        :desc "vsplit"    :n "/"  #'evil-window-vsplit
        :desc "split"     :n "\\" #'evil-window-split))
